Documentation/High Availability and Scale.md

```diff
@@ -0,0 +1,369 @@
+
+**Considerations**
+
+1. ENIs from single VM provisioned on multiple cards on same SDN Appliance
+1. Card from SDN Appliance 1 is "paired" with card from SDN Appliance 2 (each card participates only in single pairing relationship)
+1. Pairwise flow replication
+1. Single card handles multiple ENIs
+1. Some ENIs on same card are in "Active" mode, some other ENIs on same card are in "Passive" mode
+1. Each card has two VIPs: one VIP announced with short path thru BGP (used by ENIs in "Active" mode), 
+second VIP announced with longer path thru BGP (used by ENIs in "Passive" mode)
+1. Paired cards announce same set of VIPs
+1. ENI-based (not card based) flow replication (direction of flow replication: "Active ENI" -\> "Passive ENI")
+1. Single ENI is programmed on multiple cards (each card with different VIP)
+1. TOR (or source node where VM is located) performs traffic load balancing/splitting/sharding for 
+selected ENI across VIPs of the cards on which this ENI is provisioned
+
+## High Availability Architecture
+
+### Active-Passive model
+
+Active-Passive model will ensure high availability in case either (1)
+one of the TOR/SDN Appliance dies, or (2) single card dies.
+
+**Setup details**
+
+1. Each card from "SDN Appliance 1" will have "paired" card from "SDN Appliance 2"
+1. "Paired" cards will be serving same ENI with exact same policies setup on each
+1. "Paired" cards will be continuously replicating active flows from Active card to Passive card
+1.  Both cards will be announcing same VIP via BGP
+1. "SDN Appliance 1" will be in Active mode (announcing preferred, shorter path to itself thru BGP)
+1. "SDN Appliance 2" will be in Passive mode (announcing less preferred, longer path to itself thru BGP)
+
+**Normal traffic pattern**
+
+Normal traffic pattern for ENIs handled by cards will be going always thru "SDN Appliance 1" (Active one).
+
+**Failure mode**
+
+In case of failure the BGP routes from "SDN Appliance 1" (previously active) will be withdrawn and TOR will prefer 
+"SDN Appliance 2" and redirect traffic there, ensuring continuous traffic and uninterrupted customer experience.
+
+### Pairing
+
+Cards between "SDN Appliance 1" and "SDN Appliance 2" will be paired with each other to created described "Active-Passive" model.
+
+Control plane will be responsible for creating "pairing" relationship between cards -- select which cards create a pair.
+
+### ENI Policy configuration
+
+"Paired" cards will be configured (by control plane) with same ENI and same policy.
+
+Control plane will be responsible for configuring same ENI and same policy across both paired cards. 
+To clarify - no replication of ENI policy is required by card -- control plane will do this.
+
+### Flow replication
+
+Once "pairing" relationship is established, flows will need to start being replicated and synced.
+
+Cards are responsible for replicating and syncing flows across cards in "paired" relationship.
```

Comment:
How do cards "talk" to each other? 1. What is the physical path? 2. Is there supposed to be a standard protocol for doing this or is it up to the vendors to invent their own?

Reply:
@mzms - how do the cards 'talk' to each other?

Reply:
Question for @mzms.